File: app-runtime-management/agents/macos/Sources/MacOSAppRuntimeCore/AccountingV2.swift
import CryptoKit
import Foundation

public enum AccountingV2Constants {
    public static let schemaVersion = 2
    public static let idleThresholdMilliseconds: Int64 = 180_000
    public static let checkpointIntervalMilliseconds: Int64 = 60_000
    public static let estimatedGapCapMilliseconds: Int64 = 30_000
    public static let reorderWindowMilliseconds: Int64 = 500
}

public enum AccountingFactKind: String, Codable, Sendable {
    case foregroundChanged
    case userActivityChanged
    case sessionChanged
    case powerChanged
    case pipChanged
    case mediaChanged
    case checkpoint
    case clockAdjusted
    case recovery
}

public enum UsageChannel: String, Codable, Sendable {
    case active
    case pipActive
    case diagnostic
}

public enum ActivityBasis: String, Codable, Sendable {
    case foregroundInteraction
    case foregroundStrongMedia
    case pipStrongMedia
    case estimatedCheckpoint
    case estimatedBackfill
    case estimatedRecovery
    case diagnostic
}

public enum WindowPresentationState: String, Codable, Sendable {
    case unknown
    case visible
    case hidden
    case minimized
}

public enum MediaEvidenceLevel: String, Codable, Sendable {
    case none
    case weak
    case strong
}

public enum MediaPlaybackState: String, Codable, Sendable {
    case unknown
    case playing
    case paused
    case stopped
}

public enum MediaKind: String, Codable, Sendable {
    case audio
    case video
}

public enum MediaPresentation: String, Codable, Sendable {
    case foreground
    case background
    case pip
}

public enum PipState: String, Codable, Sendable {
    case inactive
    case active
}

public enum CheckpointConfirmation: String, Codable, Sendable {
    case confirmed
    case failed
}

public enum ApplicationClassification: String, Codable, Sendable {
    case study
    case composite
    case restrictedEntertainment
    case unclassified
    case blocked
}

public struct AccountingPolicySnapshot: Codable, Equatable, Sendable {
    public let assignmentVersion: Int64?
    public let appPolicyVersion: Int64?
    public let applicationClassification: ApplicationClassification?
    public let quotaBucket: String?
}

public struct AccountingRuntimeSnapshot: Codable, Equatable, Sendable {
    public let foregroundApplication: ApplicationIdentity?
    public let foregroundWindowState: WindowPresentationState
    public let foregroundMediaEvidence: MediaEvidenceLevel
    public let foregroundPlaybackState: MediaPlaybackState
    public let userActivity: UserActivityState
    public let sessionState: UserSessionState
    public let powerState: SystemPowerState
}

public struct AccountingRuntimeFact: Codable, Equatable, Sendable {
    public let wallTimeMs: Int64
    public let monotonicTimeMs: Int64
    public let clockEpochId: String
    public let kind: AccountingFactKind
    public let application: ApplicationIdentity?
    public let userActivity: UserActivityState?
    public let sessionState: UserSessionState?
    public let powerState: SystemPowerState?
    public let windowState: WindowPresentationState
    public let mediaEvidence: MediaEvidenceLevel
    public let playbackState: MediaPlaybackState
    public let pipState: PipState?
    public let mediaKind: MediaKind?
    public let mediaPresentation: MediaPresentation?
    public let confirmation: CheckpointConfirmation?
    public let snapshot: AccountingRuntimeSnapshot?
    public let newClockEpochId: String?
    public let diagnosticHint: String?

    public init(
        wallTimeMs: Int64,
        monotonicTimeMs: Int64,
        clockEpochId: String,
        kind: AccountingFactKind,
        application: ApplicationIdentity? = nil,
        userActivity: UserActivityState? = nil,
        sessionState: UserSessionState? = nil,
        powerState: SystemPowerState? = nil,
        windowState: WindowPresentationState = .unknown,
        mediaEvidence: MediaEvidenceLevel = .none,
        playbackState: MediaPlaybackState = .unknown,
        pipState: PipState? = nil,
        mediaKind: MediaKind? = nil,
        mediaPresentation: MediaPresentation? = nil,
        confirmation: CheckpointConfirmation? = nil,
        snapshot: AccountingRuntimeSnapshot? = nil,
        newClockEpochId: String? = nil,
        diagnosticHint: String? = nil
    ) {
        self.wallTimeMs = wallTimeMs
        self.monotonicTimeMs = monotonicTimeMs
        self.clockEpochId = clockEpochId
        self.kind = kind
        self.application = application
        self.userActivity = userActivity
        self.sessionState = sessionState
        self.powerState = powerState
        self.windowState = windowState
        self.mediaEvidence = mediaEvidence
        self.playbackState = playbackState
        self.pipState = pipState
        self.mediaKind = mediaKind
        self.mediaPresentation = mediaPresentation
        self.confirmation = confirmation
        self.snapshot = snapshot
        self.newClockEpochId = newClockEpochId
        self.diagnosticHint = diagnosticHint
    }

    private enum CodingKeys: String, CodingKey {
        case wallTimeMs, monotonicTimeMs, clockEpochId, kind, application
        case userActivity, sessionState, powerState, windowState, mediaEvidence
        case playbackState, pipState, mediaKind, mediaPresentation, confirmation
        case snapshot, newClockEpochId, diagnosticHint
    }

    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        wallTimeMs = try values.decode(Int64.self, forKey: .wallTimeMs)
        monotonicTimeMs = try values.decode(Int64.self, forKey: .monotonicTimeMs)
        clockEpochId = try values.decode(String.self, forKey: .clockEpochId)
        kind = try values.decode(AccountingFactKind.self, forKey: .kind)
        application = try values.decodeIfPresent(ApplicationIdentity.self, forKey: .application)
        userActivity = try values.decodeIfPresent(UserActivityState.self, forKey: .userActivity)
        sessionState = try values.decodeIfPresent(UserSessionState.self, forKey: .sessionState)
        powerState = try values.decodeIfPresent(SystemPowerState.self, forKey: .powerState)
        windowState = try values.decodeIfPresent(WindowPresentationState.self, forKey: .windowState) ?? .unknown
        mediaEvidence = try values.decodeIfPresent(MediaEvidenceLevel.self, forKey: .mediaEvidence) ?? .none
        playbackState = try values.decodeIfPresent(MediaPlaybackState.self, forKey: .playbackState) ?? .unknown
        pipState = try values.decodeIfPresent(PipState.self, forKey: .pipState)
        mediaKind = try values.decodeIfPresent(MediaKind.self, forKey: .mediaKind)
        mediaPresentation = try values.decodeIfPresent(MediaPresentation.self, forKey: .mediaPresentation)
        confirmation = try values.decodeIfPresent(CheckpointConfirmation.self, forKey: .confirmation)
        snapshot = try values.decodeIfPresent(AccountingRuntimeSnapshot.self, forKey: .snapshot)
        newClockEpochId = try values.decodeIfPresent(String.self, forKey: .newClockEpochId)
        diagnosticHint = try values.decodeIfPresent(String.self, forKey: .diagnosticHint)
    }

    public var safetyPriority: Int {
        switch kind {
        case .clockAdjusted: return 0
        case .sessionChanged where sessionState != .active: return 1
        case .powerChanged where powerState != .awake: return 2
        case .userActivityChanged where userActivity == .idle: return 3
        case .pipChanged where pipState == .inactive: return 4
        case .mediaChanged where playbackState != .playing: return 5
        case .foregroundChanged: return 10
        case .userActivityChanged: return 11
        case .sessionChanged: return 12
        case .powerChanged: return 13
        case .pipChanged: return 14
        case .mediaChanged: return 15
        case .recovery: return 20
        case .checkpoint: return 30
        default: return 40
        }
    }
}

public struct EstimatedMetadata: Codable, Equatable, Sendable {
    public let isEstimated: Bool
    public let reason: String?
    public let cappedAtMilliseconds: Int64?

    public static let exact = EstimatedMetadata(
        isEstimated: false,
        reason: nil,
        cappedAtMilliseconds: nil
    )
}

public struct UsageSegmentV2: Codable, Equatable, Sendable {
    public let id: String
    public let schemaVersion: Int
    public let runtimeSessionID: String
    public let application: ApplicationIdentity?
    public let channel: UsageChannel
    public let activityBasis: ActivityBasis
    public let clockEpochId: String
    public let startWallTimeMs: Int64
    public let endWallTimeMs: Int64
    public let startMonotonicTimeMs: Int64
    public let endMonotonicTimeMs: Int64
    public let monotonicDurationMilliseconds: Int64
    public let endReason: SegmentEndReason
    public let estimated: EstimatedMetadata
    public let lastEvidenceWallTimeMs: Int64?
    public let lastEvidenceMonotonicTimeMs: Int64?
    public let diagnostic: Bool
    public let diagnosticCode: String?
    public let diagnosticMessage: String?
    public let policySnapshot: AccountingPolicySnapshot?

    public var authoritativeForUsage: Bool { !diagnostic && channel != .diagnostic }

    public static func create(
        runtimeSessionID: String,
        application: ApplicationIdentity?,
        channel requestedChannel: UsageChannel,
        activityBasis requestedBasis: ActivityBasis,
        clockEpochId: String,
        startWallTimeMs: Int64,
        endWallTimeMs: Int64,
        startMonotonicTimeMs: Int64,
        endMonotonicTimeMs: Int64,
        endReason requestedEndReason: SegmentEndReason,
        estimated requestedEstimated: EstimatedMetadata,
        lastEvidenceWallTimeMs: Int64?,
        lastEvidenceMonotonicTimeMs: Int64?,
        diagnostic requestedDiagnostic: Bool = false,
        diagnosticCode requestedDiagnosticCode: String? = nil,
        diagnosticMessage requestedDiagnosticMessage: String? = nil,
        policySnapshot: AccountingPolicySnapshot? = nil
    ) -> UsageSegmentV2 {
        let duration = max(0, endMonotonicTimeMs - startMonotonicTimeMs)
        let zeroBoundary = duration == 0 && !requestedDiagnostic
        let channel = zeroBoundary ? UsageChannel.diagnostic : requestedChannel
        let basis = zeroBoundary ? ActivityBasis.diagnostic : requestedBasis
        let endReason = zeroBoundary ? SegmentEndReason.diagnostic : requestedEndReason
        let estimated = zeroBoundary ? EstimatedMetadata.exact : requestedEstimated
        let diagnostic = zeroBoundary || requestedDiagnostic
        let diagnosticCode = zeroBoundary ? "zeroDurationBoundary" : requestedDiagnosticCode
        let diagnosticMessage = zeroBoundary
            ? (requestedDiagnosticMessage ?? "A same-millisecond boundary did not produce billable duration.")
            : requestedDiagnosticMessage
        let canonical = AccountingSegmentID.canonicalUsage(
            runtimeSessionID: runtimeSessionID,
            application: application,
            channel: channel,
            basis: basis,
            clockEpochId: clockEpochId,
            startWall: startWallTimeMs,
            endWall: endWallTimeMs,
            startMonotonic: startMonotonicTimeMs,
            endMonotonic: endMonotonicTimeMs,
            duration: duration,
            endReason: endReason,
            estimated: estimated,
            diagnostic: diagnostic,
            diagnosticCode: diagnosticCode
        )
        return UsageSegmentV2(
            id: AccountingSegmentID.sha256(canonical),
            schemaVersion: AccountingV2Constants.schemaVersion,
            runtimeSessionID: runtimeSessionID,
            application: application,
            channel: channel,
            activityBasis: basis,
            clockEpochId: clockEpochId,
            startWallTimeMs: startWallTimeMs,
            endWallTimeMs: endWallTimeMs,
            startMonotonicTimeMs: startMonotonicTimeMs,
            endMonotonicTimeMs: endMonotonicTimeMs,
            monotonicDurationMilliseconds: duration,
            endReason: endReason,
            estimated: estimated,
            lastEvidenceWallTimeMs: lastEvidenceWallTimeMs,
            lastEvidenceMonotonicTimeMs: lastEvidenceMonotonicTimeMs,
            diagnostic: diagnostic,
            diagnosticCode: diagnosticCode,
            diagnosticMessage: diagnosticMessage,
            policySnapshot: policySnapshot
        )
    }
}

public struct MediaSegmentV2: Codable, Equatable, Sendable {
    public let id: String
    public let schemaVersion: Int
    public let runtimeSessionID: String
    public let application: ApplicationIdentity
    public let mediaKind: MediaKind
    public let presentation: MediaPresentation
    public let clockEpochId: String
    public let startWallTimeMs: Int64
    public let endWallTimeMs: Int64
    public let startMonotonicTimeMs: Int64
    public let endMonotonicTimeMs: Int64
    public let monotonicDurationMilliseconds: Int64
    public let endReason: SegmentEndReason
    public let estimated: EstimatedMetadata
    public let lastEvidenceWallTimeMs: Int64
    public let lastEvidenceMonotonicTimeMs: Int64
    public let authoritativeForUsage: Bool

    public static func create(
        runtimeSessionID: String,
        application: ApplicationIdentity,
        mediaKind: MediaKind,
        presentation: MediaPresentation,
        clockEpochId: String,
        startWallTimeMs: Int64,
        endWallTimeMs: Int64,
        startMonotonicTimeMs: Int64,
        endMonotonicTimeMs: Int64,
        endReason: SegmentEndReason,
        estimated: EstimatedMetadata,
        lastEvidenceWallTimeMs: Int64,
        lastEvidenceMonotonicTimeMs: Int64
    ) -> MediaSegmentV2 {
        let duration = max(0, endMonotonicTimeMs - startMonotonicTimeMs)
        let canonical = AccountingSegmentID.canonicalMedia(
            runtimeSessionID: runtimeSessionID,
            application: application,
            mediaKind: mediaKind,
            presentation: presentation,
            clockEpochId: clockEpochId,
            startWall: startWallTimeMs,
            endWall: endWallTimeMs,
            startMonotonic: startMonotonicTimeMs,
            endMonotonic: endMonotonicTimeMs,
            duration: duration,
            endReason: endReason,
            estimated: estimated
        )
        return MediaSegmentV2(
            id: AccountingSegmentID.sha256(canonical),
            schemaVersion: AccountingV2Constants.schemaVersion,
            runtimeSessionID: runtimeSessionID,
            application: application,
            mediaKind: mediaKind,
            presentation: presentation,
            clockEpochId: clockEpochId,
            startWallTimeMs: startWallTimeMs,
            endWallTimeMs: endWallTimeMs,
            startMonotonicTimeMs: startMonotonicTimeMs,
            endMonotonicTimeMs: endMonotonicTimeMs,
            monotonicDurationMilliseconds: duration,
            endReason: endReason,
            estimated: estimated,
            lastEvidenceWallTimeMs: lastEvidenceWallTimeMs,
            lastEvidenceMonotonicTimeMs: lastEvidenceMonotonicTimeMs,
            authoritativeForUsage: false
        )
    }
}

public enum AccountingSegmentID {
    public static func sha256(_ canonical: String) -> String {
        SHA256.hash(data: Data(canonical.utf8)).map { String(format: "%02x", $0) }.joined()
    }

    public static func canonicalUsage(
        runtimeSessionID: String,
        application: ApplicationIdentity?,
        channel: UsageChannel,
        basis: ActivityBasis,
        clockEpochId: String,
        startWall: Int64,
        endWall: Int64,
        startMonotonic: Int64,
        endMonotonic: Int64,
        duration: Int64,
        endReason: SegmentEndReason,
        estimated: EstimatedMetadata,
        diagnostic: Bool,
        diagnosticCode: String?
    ) -> String {
        [
            "usage-v2", runtimeSessionID, application?.platform.rawValue ?? "",
            application?.runtimeIdentity ?? "", channel.rawValue, basis.rawValue,
            clockEpochId, String(startWall), String(endWall), String(startMonotonic),
            String(endMonotonic), String(duration), endReason.rawValue,
            estimated.isEstimated ? "1" : "0", estimated.reason ?? "",
            estimated.cappedAtMilliseconds.map { String($0) } ?? "",
            diagnostic ? "1" : "0", diagnosticCode ?? ""
        ].joined(separator: "\n")
    }

    public static func canonicalMedia(
        runtimeSessionID: String,
        application: ApplicationIdentity,
        mediaKind: MediaKind,
        presentation: MediaPresentation,
        clockEpochId: String,
        startWall: Int64,
        endWall: Int64,
        startMonotonic: Int64,
        endMonotonic: Int64,
        duration: Int64,
        endReason: SegmentEndReason,
        estimated: EstimatedMetadata
    ) -> String {
        [
            "media-v2", runtimeSessionID, application.platform.rawValue,
            application.runtimeIdentity, mediaKind.rawValue, presentation.rawValue,
            clockEpochId, String(startWall), String(endWall), String(startMonotonic),
            String(endMonotonic), String(duration), endReason.rawValue,
            estimated.isEstimated ? "1" : "0", estimated.reason ?? "",
            estimated.cappedAtMilliseconds.map { String($0) } ?? ""
        ].joined(separator: "\n")
    }
}

public struct OpenAccountingLane: Equatable, Codable, Sendable {
    public let application: ApplicationIdentity
    public let channel: UsageChannel
    public let activityBasis: ActivityBasis
    public let clockEpochId: String
    public let startWallTimeMs: Int64
    public let startMonotonicTimeMs: Int64
    public let lastEvidenceWallTimeMs: Int64
    public let lastEvidenceMonotonicTimeMs: Int64
}

public struct OpenMediaLane: Equatable, Codable, Sendable {
    public let application: ApplicationIdentity
    public let mediaKind: MediaKind
    public let presentation: MediaPresentation
    public let clockEpochId: String
    public let startWallTimeMs: Int64
    public let startMonotonicTimeMs: Int64
    public let lastEvidenceWallTimeMs: Int64
    public let lastEvidenceMonotonicTimeMs: Int64
}

public struct PipObservation: Equatable, Codable, Sendable {
    public let application: ApplicationIdentity
    public let windowState: WindowPresentationState
    public let mediaEvidence: MediaEvidenceLevel
    public let playbackState: MediaPlaybackState
}

public struct MediaObservation: Equatable, Codable, Sendable {
    public let application: ApplicationIdentity
    public let mediaKind: MediaKind
    public let presentation: MediaPresentation
    public let mediaEvidence: MediaEvidenceLevel
    public let playbackState: MediaPlaybackState
}

public struct AccountingRuntimeState: Equatable, Codable, Sendable {
    public let runtimeSessionID: String
    public var clockEpochId: String
    public var foregroundApplication: ApplicationIdentity?
    public var foregroundWindowState: WindowPresentationState = .unknown
    public var foregroundMediaEvidence: MediaEvidenceLevel = .none
    public var foregroundPlaybackState: MediaPlaybackState = .unknown
    public var userActivity: UserActivityState = .unknown
    public var sessionState: UserSessionState = .unknown
    public var powerState: SystemPowerState = .unknown
    public var foregroundLane: OpenAccountingLane?
    public var pipObservations: [String: PipObservation] = [:]
    public var pipLanes: [String: OpenAccountingLane] = [:]
    public var mediaObservations: [String: MediaObservation] = [:]
    public var mediaLanes: [String: OpenMediaLane] = [:]
    public var lastProcessedWallTimeMs: Int64?
    public var lastProcessedMonotonicTimeMs: Int64?

    public init(runtimeSessionID: String, clockEpochId: String = "epoch-0") {
        self.runtimeSessionID = runtimeSessionID
        self.clockEpochId = clockEpochId
    }
}

public struct AccountingTransition: Equatable, Sendable {
    public let state: AccountingRuntimeState
    public let usageSegments: [UsageSegmentV2]
    public let mediaSegments: [MediaSegmentV2]
}

File: app-runtime-management/agents/windows/src/TimeOnChrome.AppRuntime.Core/AccountingV2.cs
using System.Security.Cryptography;
using System.Text;

namespace TimeOnChrome.AppRuntime.Core;

public static class AccountingV2Constants
{
    public const int SchemaVersion = 2;
    public const long IdleThresholdMilliseconds = 180_000;
    public const long CheckpointIntervalMilliseconds = 60_000;
    public const long EstimatedGapCapMilliseconds = 30_000;
    public const long ReorderWindowMilliseconds = 500;
}

public enum AccountingFactKind
{
    ForegroundChanged,
    UserActivityChanged,
    SessionChanged,
    PowerChanged,
    PipChanged,
    MediaChanged,
    Checkpoint,
    ClockAdjusted,
    Recovery,
}

public enum UsageChannel
{
    Active,
    PipActive,
    Diagnostic,
}

public enum ActivityBasis
{
    ForegroundInteraction,
    ForegroundStrongMedia,
    PipStrongMedia,
    EstimatedCheckpoint,
    EstimatedBackfill,
    EstimatedRecovery,
    Diagnostic,
}

public enum WindowPresentationState
{
    Unknown,
    Visible,
    Hidden,
    Minimized,
}

public enum MediaEvidenceLevel
{
    None,
    Weak,
    Strong,
}

public enum MediaPlaybackState
{
    Unknown,
    Playing,
    Paused,
    Stopped,
}

public enum MediaKind
{
    Audio,
    Video,
}

public enum MediaPresentation
{
    Foreground,
    Background,
    Pip,
}

public enum PipState
{
    Inactive,
    Active,
}

public enum CheckpointConfirmation
{
    Confirmed,
    Failed,
}

public enum ApplicationClassification
{
    Study,
    Composite,
    RestrictedEntertainment,
    Unclassified,
    Blocked,
}

public sealed record AccountingPolicySnapshot(
    long? AssignmentVersion,
    string? QuotaBucket,
    long? AppPolicyVersion = null,
    ApplicationClassification? ApplicationClassification = null);

public sealed record AccountingRuntimeSnapshot(
    ApplicationIdentity? ForegroundApplication,
    WindowPresentationState ForegroundWindowState,
    MediaEvidenceLevel ForegroundMediaEvidence,
    MediaPlaybackState ForegroundPlaybackState,
    UserActivityState UserActivity,
    UserSessionState SessionState,
    SystemPowerState PowerState);

public sealed record AccountingRuntimeFact(
    long WallTimeMs,
    long MonotonicTimeMs,
    string ClockEpochId,
    AccountingFactKind Kind,
    ApplicationIdentity? Application = null,
    UserActivityState? UserActivity = null,
    UserSessionState? SessionState = null,
    SystemPowerState? PowerState = null,
    WindowPresentationState WindowState = WindowPresentationState.Unknown,
    MediaEvidenceLevel MediaEvidence = MediaEvidenceLevel.None,
    MediaPlaybackState PlaybackState = MediaPlaybackState.Unknown,
    PipState? PipState = null,
    MediaKind? MediaKind = null,
    MediaPresentation? MediaPresentation = null,
    CheckpointConfirmation? Confirmation = null,
    AccountingRuntimeSnapshot? Snapshot = null,
    string? NewClockEpochId = null,
    string? DiagnosticHint = null,
    AccountingPolicySnapshot? PolicySnapshot = null)
{
    public int SafetyPriority => Kind switch
    {
        AccountingFactKind.ClockAdjusted => 0,
        AccountingFactKind.SessionChanged when SessionState != UserSessionState.Active => 1,
        AccountingFactKind.PowerChanged when PowerState != SystemPowerState.Awake => 2,
        AccountingFactKind.UserActivityChanged when UserActivity == UserActivityState.Idle => 3,
        AccountingFactKind.PipChanged when PipState == TimeOnChrome.AppRuntime.Core.PipState.Inactive => 4,
        AccountingFactKind.MediaChanged when PlaybackState != MediaPlaybackState.Playing => 5,
        AccountingFactKind.ForegroundChanged => 10,
        AccountingFactKind.UserActivityChanged => 11,
        AccountingFactKind.SessionChanged => 12,
        AccountingFactKind.PowerChanged => 13,
        AccountingFactKind.PipChanged => 14,
        AccountingFactKind.MediaChanged => 15,
        AccountingFactKind.Recovery => 20,
        AccountingFactKind.Checkpoint => 30,
        _ => 40,
    };
}

public sealed record EstimatedMetadata(bool IsEstimated, string? Reason, long? CappedAtMilliseconds)
{
    public static EstimatedMetadata Exact { get; } = new(false, null, null);
}

public sealed record UsageSegmentV2(
    string Id,
    int SchemaVersion,
    string RuntimeSessionID,
    ApplicationIdentity? Application,
    UsageChannel Channel,
    ActivityBasis ActivityBasis,
    string ClockEpochId,
    long StartWallTimeMs,
    long EndWallTimeMs,
    long StartMonotonicTimeMs,
    long EndMonotonicTimeMs,
    long MonotonicDurationMilliseconds,
    SegmentEndReason EndReason,
    EstimatedMetadata Estimated,
    long? LastEvidenceWallTimeMs,
    long? LastEvidenceMonotonicTimeMs,
    bool Diagnostic,
    string? DiagnosticCode,
    string? DiagnosticMessage,
    AccountingPolicySnapshot? PolicySnapshot)
{
    public bool AuthoritativeForUsage => !Diagnostic && Channel != UsageChannel.Diagnostic;

    public static UsageSegmentV2 Create(
        string runtimeSessionID,
        ApplicationIdentity? application,
        UsageChannel channel,
        ActivityBasis activityBasis,
        string clockEpochId,
        long startWallTimeMs,
        long endWallTimeMs,
        long startMonotonicTimeMs,
        long endMonotonicTimeMs,
        SegmentEndReason endReason,
        EstimatedMetadata estimated,
        long? lastEvidenceWallTimeMs,
        long? lastEvidenceMonotonicTimeMs,
        bool diagnostic = false,
        string? diagnosticCode = null,
        string? diagnosticMessage = null,
        AccountingPolicySnapshot? policySnapshot = null)
    {
        var duration = Math.Max(0, endMonotonicTimeMs - startMonotonicTimeMs);
        if (duration == 0 && !diagnostic)
        {
            channel = UsageChannel.Diagnostic;
            activityBasis = ActivityBasis.Diagnostic;
            endReason = SegmentEndReason.Diagnostic;
            estimated = EstimatedMetadata.Exact;
            diagnostic = true;
            diagnosticCode = "zeroDurationBoundary";
            diagnosticMessage ??= "A same-millisecond boundary did not produce billable duration.";
        }
        var canonical = AccountingSegmentId.CanonicalUsage(
            runtimeSessionID,
            application,
            channel,
            activityBasis,
            clockEpochId,
            startWallTimeMs,
            endWallTimeMs,
            startMonotonicTimeMs,
            endMonotonicTimeMs,
            duration,
            endReason,
            estimated,
            diagnostic,
            diagnosticCode);
        return new UsageSegmentV2(
            AccountingSegmentId.Sha256(canonical),
            AccountingV2Constants.SchemaVersion,
            runtimeSessionID,
            application,
            channel,
            activityBasis,
            clockEpochId,
            startWallTimeMs,
            endWallTimeMs,
            startMonotonicTimeMs,
            endMonotonicTimeMs,
            duration,
            endReason,
            estimated,
            lastEvidenceWallTimeMs,
            lastEvidenceMonotonicTimeMs,
            diagnostic,
            diagnosticCode,
            diagnosticMessage,
            policySnapshot);
    }
}

public sealed record MediaSegmentV2(
    string Id,
    int SchemaVersion,
    string RuntimeSessionID,
    ApplicationIdentity Application,
    MediaKind MediaKind,
    MediaPresentation Presentation,
    string ClockEpochId,
    long StartWallTimeMs,
    long EndWallTimeMs,
    long StartMonotonicTimeMs,
    long EndMonotonicTimeMs,
    long MonotonicDurationMilliseconds,
    SegmentEndReason EndReason,
    EstimatedMetadata Estimated,
    long LastEvidenceWallTimeMs,
    long LastEvidenceMonotonicTimeMs,
    bool AuthoritativeForUsage = false)
{
    public static MediaSegmentV2 Create(
        string runtimeSessionID,
        ApplicationIdentity application,
        MediaKind mediaKind,
        MediaPresentation presentation,
        string clockEpochId,
        long startWallTimeMs,
        long endWallTimeMs,
        long startMonotonicTimeMs,
        long endMonotonicTimeMs,
        SegmentEndReason endReason,
        EstimatedMetadata estimated,
        long lastEvidenceWallTimeMs,
        long lastEvidenceMonotonicTimeMs)
    {
        var duration = Math.Max(0, endMonotonicTimeMs - startMonotonicTimeMs);
        var canonical = AccountingSegmentId.CanonicalMedia(
            runtimeSessionID,
            application,
            mediaKind,
            presentation,
            clockEpochId,
            startWallTimeMs,
            endWallTimeMs,
            startMonotonicTimeMs,
            endMonotonicTimeMs,
            duration,
            endReason,
            estimated);
        return new MediaSegmentV2(
            AccountingSegmentId.Sha256(canonical),
            AccountingV2Constants.SchemaVersion,
            runtimeSessionID,
            application,
            mediaKind,
            presentation,
            clockEpochId,
            startWallTimeMs,
            endWallTimeMs,
            startMonotonicTimeMs,
            endMonotonicTimeMs,
            duration,
            endReason,
            estimated,
            lastEvidenceWallTimeMs,
            lastEvidenceMonotonicTimeMs,
            false);
    }
}

public static class AccountingSegmentId
{
    public static string Sha256(string canonical)
    {
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(canonical))).ToLowerInvariant();
    }

    public static string CanonicalUsage(
        string runtimeSessionID,
        ApplicationIdentity? application,
        UsageChannel channel,
        ActivityBasis basis,
        string clockEpochId,
        long startWall,
        long endWall,
        long startMonotonic,
        long endMonotonic,
        long duration,
        SegmentEndReason endReason,
        EstimatedMetadata estimated,
        bool diagnostic,
        string? diagnosticCode)
    {
        return string.Join('\n', new[]
        {
            "usage-v2",
            runtimeSessionID,
            application?.Platform.ToString().ToLowerInvariant() ?? string.Empty,
            application?.RuntimeIdentity ?? string.Empty,
            Wire(channel),
            Wire(basis),
            clockEpochId,
            startWall.ToString(System.Globalization.CultureInfo.InvariantCulture),
            endWall.ToString(System.Globalization.CultureInfo.InvariantCulture),
            startMonotonic.ToString(System.Globalization.CultureInfo.InvariantCulture),
            endMonotonic.ToString(System.Globalization.CultureInfo.InvariantCulture),
            duration.ToString(System.Globalization.CultureInfo.InvariantCulture),
            Wire(endReason),
            estimated.IsEstimated ? "1" : "0",
            estimated.Reason ?? string.Empty,
            estimated.CappedAtMilliseconds?.ToString(System.Globalization.CultureInfo.InvariantCulture) ?? string.Empty,
            diagnostic ? "1" : "0",
            diagnosticCode ?? string.Empty,
        });
    }

    public static string CanonicalMedia(
        string runtimeSessionID,
        ApplicationIdentity application,
        MediaKind kind,
        MediaPresentation presentation,
        string clockEpochId,
        long startWall,
        long endWall,
        long startMonotonic,
        long endMonotonic,
        long duration,
        SegmentEndReason endReason,
        EstimatedMetadata estimated)
    {
        return string.Join('\n', new[]
        {
            "media-v2",
            runtimeSessionID,
            application.Platform.ToString().ToLowerInvariant(),
            application.RuntimeIdentity,
            Wire(kind),
            Wire(presentation),
            clockEpochId,
            startWall.ToString(System.Globalization.CultureInfo.InvariantCulture),
            endWall.ToString(System.Globalization.CultureInfo.InvariantCulture),
            startMonotonic.ToString(System.Globalization.CultureInfo.InvariantCulture),
            endMonotonic.ToString(System.Globalization.CultureInfo.InvariantCulture),
            duration.ToString(System.Globalization.CultureInfo.InvariantCulture),
            Wire(endReason),
            estimated.IsEstimated ? "1" : "0",
            estimated.Reason ?? string.Empty,
            estimated.CappedAtMilliseconds?.ToString(System.Globalization.CultureInfo.InvariantCulture) ?? string.Empty,
        });
    }

    public static string Wire<T>(T value) where T : struct, Enum
    {
        var text = value.ToString();
        return char.ToLowerInvariant(text[0]) + text[1..];
    }
}

public sealed record OpenAccountingLane(
    ApplicationIdentity Application,
    UsageChannel Channel,
    ActivityBasis ActivityBasis,
    string ClockEpochId,
    long StartWallTimeMs,
    long StartMonotonicTimeMs,
    long LastEvidenceWallTimeMs,
    long LastEvidenceMonotonicTimeMs,
    AccountingPolicySnapshot? PolicySnapshot = null);

public sealed record OpenMediaLane(
    ApplicationIdentity Application,
    MediaKind MediaKind,
    MediaPresentation Presentation,
    string ClockEpochId,
    long StartWallTimeMs,
    long StartMonotonicTimeMs,
    long LastEvidenceWallTimeMs,
    long LastEvidenceMonotonicTimeMs);

public sealed record PipObservation(
    ApplicationIdentity Application,
    WindowPresentationState WindowState,
    MediaEvidenceLevel MediaEvidence,
    MediaPlaybackState PlaybackState);

public sealed record MediaObservation(
    ApplicationIdentity Application,
    MediaKind MediaKind,
    MediaPresentation Presentation,
    MediaEvidenceLevel MediaEvidence,
    MediaPlaybackState PlaybackState);

public sealed record AccountingTransition(
    AccountingRuntimeState State,
    IReadOnlyList<UsageSegmentV2> UsageSegments,
    IReadOnlyList<MediaSegmentV2> MediaSegments);

public sealed record AccountingRuntimeState
{
    public required string RuntimeSessionID { get; init; }
    public string ClockEpochId { get; init; } = "epoch-0";
    public ApplicationIdentity? ForegroundApplication { get; init; }
    public WindowPresentationState ForegroundWindowState { get; init; } = WindowPresentationState.Unknown;
    public MediaEvidenceLevel ForegroundMediaEvidence { get; init; } = MediaEvidenceLevel.None;
    public MediaPlaybackState ForegroundPlaybackState { get; init; } = MediaPlaybackState.Unknown;
    public UserActivityState UserActivity { get; init; } = UserActivityState.Unknown;
    public UserSessionState SessionState { get; init; } = UserSessionState.Unknown;
    public SystemPowerState PowerState { get; init; } = SystemPowerState.Unknown;
    public OpenAccountingLane? ForegroundLane { get; init; }
    public IReadOnlyDictionary<string, PipObservation> PipObservations { get; init; } = new Dictionary<string, PipObservation>();
    public IReadOnlyDictionary<string, OpenAccountingLane> PipLanes { get; init; } = new Dictionary<string, OpenAccountingLane>();
    public IReadOnlyDictionary<string, MediaObservation> MediaObservations { get; init; } = new Dictionary<string, MediaObservation>();
    public IReadOnlyDictionary<string, OpenMediaLane> MediaLanes { get; init; } = new Dictionary<string, OpenMediaLane>();
    public long? LastProcessedWallTimeMs { get; init; }
    public long? LastProcessedMonotonicTimeMs { get; init; }
}
